cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs

```diff
@@ -0,0 +1,201 @@
+      , "useLedgerAfterSlot" .= useLedger ns
+      ]
+
+data RootAddress = RootAddress
+  { addrs :: [RelayAccessPoint]
+  , advertise :: PeerAdvertise
+  } deriving (Eq, Show)
+
+instance FromJSON RootAddress where
+  parseJSON = withObject "RootAddress" $ \o ->
+                RootAddress
+                  <$> o .:  "addrs"
+                  <*> o .:? "advertise" .!= DoNotAdvertisePeer
+
+instance ToJSON RootAddress where
+  toJSON ra =
+    object
+      [ "addrs" .= addrs ra
+      , "advertise" .= advertise ra
+      ]
+
+-- | Transforms a 'RootAddress' into a pair of 'RelayAccessPoint' and its
+-- corresponding 'PeerAdvertise' value.
+--
+rootAddressToRelayAccessPoint
+  :: RootAddress
+  -> [(RelayAccessPoint, PeerAdvertise)]
+rootAddressToRelayAccessPoint RootAddress { addrs, advertise } =
+    [ (ra, advertise) | ra <- addrs ]
+
+data LocalRootPeers = LocalRootPeers
+  { localRoots :: RootAddress
+  , valency :: Int
+  } deriving (Eq, Show)
+
+instance FromJSON LocalRootPeers where
+  parseJSON = withObject "LocalRootPeers" $ \o ->
+                LocalRootPeers
+                  <$> o .: "localRoots"
+                  <*> o .: "valency"
+
+instance ToJSON LocalRootPeers where
+  toJSON lrpg =
+    object
+      [ "localRoots" .= localRoots lrpg
+      , "valency" .= valency lrpg
+      ]
+
+newtype LocalRootPeersGroups = LocalRootPeersGroups
+  { groups :: [LocalRootPeers]
+  } deriving (Eq, Show)
+
+instance FromJSON LocalRootPeersGroups where
+  parseJSON = withObject "LocalRootPeersGroups" $ \o ->
+                LocalRootPeersGroups
+                  <$> o .: "groups"
+
+instance ToJSON LocalRootPeersGroups where
+  toJSON lrpg =
+    object
+      [ "groups" .= groups lrpg
+      ]
+
+newtype PublicRootPeers = PublicRootPeers
+  { publicRoots :: RootAddress
+  } deriving (Eq, Show)
+
+instance FromJSON PublicRootPeers where
+  parseJSON = withObject "PublicRootPeers" $ \o ->
+                PublicRootPeers
+                  <$> o .: "publicRoots"
+
+instance ToJSON PublicRootPeers where
+  toJSON prp =
+    object
+      [ "publicRoots" .= publicRoots prp
+      ]
+
+data NetworkTopology = RealNodeTopology !LocalRootPeersGroups ![PublicRootPeers] !UseLedger
+  deriving (Eq, Show)
+
+instance FromJSON NetworkTopology where
+  parseJSON = withObject "NetworkTopology" $ \o ->
+                RealNodeTopology <$> (o .: "LocalRoots"                                     )
+                                 <*> (o .: "PublicRoots"                                    )
+                                 <*> (o .:? "useLedgerAfterSlot" .!= UseLedger DontUseLedger)
+
+instance ToJSON NetworkTopology where
+  toJSON top =
+    case top of
+      RealNodeTopology lrpg prp ul -> object [ "LocalRoots"         .= lrpg
+                                             , "PublicRoots"        .= prp
+                                             , "useLedgerAfterSlot" .= ul
+                                             ]
```

Comment:
What's happening here where sometimes we use `PublicRoots` and sometimes `publicRoots`?

Reply:
The difference becomes easier to grasp with an example (which is pushed to documentation):

```json
{
  "LocalRoots": {
    "groups": [
      {
        "localRoots": {
          "addrs": [
            {
              "addr": "x.x.x.x",
              "port": 3001
            }
          ],
          "advertise": false
        },
        "valency": 1
      }
    ]
  },
  "PublicRoots": [
    {
      "publicRoots" : {
        "addrs": [
          {
            "addr": "y.y.y.y",
            "port": 3002
          }
        ],
        "advertise": false
      }
    }
  ],
  "useLedgerAfterSlot": 0
}
```
`PublicRoots` is the top level record name, which has various fields, including `publicRoots`; the same for `LocalRoots` and its field `localRoots`.